### PR TITLE
[SHELL32] Notify filesystem changes

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Notify filesystem change notifications by using `ReadDirectoryChangesW` function. Creating/Deleting files/folders will be responsive in Explorer.

JIRA issue: [CORE-13950](https://jira.reactos.org/browse/CORE-13950)

after

### CODE REVIEW

**Comment by yagoulas** [comment]
I think we did this in the past. Was it removed?

**Comment by katahiromz** [comment]
Yes, I deleted. You can find it here:
https://github.com/katahiromz/reactos/blob/3c5b10f4e6f0be14863b838952126c39a53a53ab/dll/win32/shell32/wine/changenotify.c#L618

**Comment by yagoulas** [comment]
Yes you deleted it in the PR where I asked you to wait for my review but it was so urgent and you didn't wait. What was so urgent anyway?

**Comment by katahiromz** [comment]
My mind is a human mind emulation, so I sometimes make irrational decisions.

**Comment by yagoulas** [comment]
I would prefer to use the previous implementation that used QueueUserAPC to receive the events. Please bring back as much code as possible from the previous implementation.

**Comment by katahiromz** [comment]
Our `QueueUserAPC` function is broken. I cannot write the fix.

**Comment by sdever** [comment]
@katahiromz I think it would be better to investigate what happens with QueueUserAPC and fix this function father than trying to workaround it so.

**Comment by yagoulas** [comment]
> 
> 
> Our `QueueUserAPC` function is broken. I cannot write the fix.

It was working before you removed it

**Comment by Saibamen** [comment]
Please mark this Pull Request as Draft: https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request

**Comment by katahiromz** [comment]
I wrote a testcase for `QueryUserAPC`: #2681

**Comment by katahiromz** [comment]
apc-failed
The current APC implementation causes fatal error...

**Comment by yagoulas** [comment]
It works now, doesn't it? With some improvements it could be merged...

**Comment by katahiromz** [comment]
Not working yet. Can you wait a few days?

**Comment by yagoulas** [comment]
I'm not in a rash, take your time.

File: dll/win32/shell32/shelldesktop/CChangeNotifyServer.cpp [hunk 1]
```diff
@@ -6,18 +6,502 @@
...
+        delete pList;
+        return NULL;
+    }
+
+    pList = AddItem(pList, szPath, TRUE);
+
+    WIN32_FIND_DATAW find;
+    PathAppendW(szPath, L"*");
+
+    HANDLE hFind = FindFirstFileW(szPath, &find);
```
Comment (yagoulas) [inline dll/win32/shell32/shelldesktop/CChangeNotifyServer.cpp:161]: What is the purpose of this?

Comment (katahiromz) [reply dll/win32/shell32/shelldesktop/CChangeNotifyServer.cpp:170]: To remember the file list of the directory.

**Comment by katahiromz** [comment]
works
Ready to go!

**Comment by katahiromz** [comment]
valid
OK.

**Comment by katahiromz** [comment]
@yagoulas Can you review this PR?

**Comment by yagoulas** [comment]
I see you didn't wait for a review once again. Now I wonder what feature you removed again, what you broke again or generally what code you made unreadable again.

**Comment by yagoulas** [comment]
Why is it so fucking hard for you to WAIT?

**Comment by ale5000-git** [comment]
It is only my opinion but personally I prefer code broken for some time rather then millions of PR waiting years to be merged and some being lost because the author is dead in the meantime :)

**Comment by ColinFinck** [comment]
@katahiromz: Asking @yagoulas for a review and then merging anyway 7 hours later is unacceptable! We already told you that about a month ago, but it seems that we weren't clear enough.

I know that our PR reviews aren't perfect and waiting can be a frustrating experience, but 7 hours don't even cover the time zone difference. Give people at least 24 hours to react after asking to review the latest changes. And reviews may take longer than that, but the time should very well be worth it as we all get better code.

Don't forget that there aren't many ReactOS shell developers.
If those few people can't even work together, all efforts are in vain..

**Comment by Mouvedia** [comment]
Did it fix CORE-16961?

**Pull request merged**